// @secretlint/secretlint-rule-no-k8s-kind-secret/module/index.js
import path from "node:path";
import { loadAll } from "js-yaml";
export const messages = {
    disallowToUseKindSecret: {
        en: (props) => `disallow to use Kind: Secret in manifest: ${props.FILE_NAME}`,
        ja: (props) => `Kind: Secretのmanifestがみつかりました: ${props.FILE_NAME}`,
    },
};
function reportIfFoundKindSecret({ source, context, t, }) {
    try {
        // Support multi manifest
        const manifestObjects = loadAll(source.content);
        manifestObjects.forEach((manifestObject) => {
            if (typeof manifestObject !== "object" || manifestObject === null) {
                return;
            }
            if (!("Kind" in manifestObject)) {
                return;
            }
            // Kind: Secret
            if (manifestObject["Kind"] !== "Secret") {
                return;
            }
            context.report({
                message: t("disallowToUseKindSecret", {
                    FILE_NAME: source.filePath ? path.basename(source.filePath) : "",
                }),
                range: [0, source.content.length],
            });
        });
    }
    catch {
        // nope
    }
}
export const creator = {
    messages,
    meta: {
        id: "@secretlint/secretlint-rule-no-k8s-kind-secret",
        recommended: false,
        type: "scanner",
        supportedContentTypes: ["text"],
        docs: {
            url: "https://github.com/secretlint/secretlint/blob/master/packages/%40secretlint/secretlint-rule-no-k8s-kind-secret/README.md",
        },
    },
    create(context, options) {
        const t = context.createTranslator(messages);
        return {
            file(source) {
                if (source.ext === ".yaml" || source.ext === ".yml") {
                    reportIfFoundKindSecret({ source, options, context, t });
                }
            },
        };
    },
};
