// @secretlint/secretlint-rule-no-homedir/module/index.js
import { matchPatterns } from "@textlint/regexp-string-matcher";
import escapeStringRegexp from "escape-string-regexp";
import os from "node:os";
const multiplatformPath = (dirPath) => {
    return new RegExp(dirPath
        .split(/[\/¥]/)
        .map((pathinfo) => escapeStringRegexp(pathinfo))
        .join("[/¥]"), "g");
};
export const messages = {
    HOMEDIR: {
        en: (props) => `found user's homedir path: ${props.PATH}`,
        ja: (props) => `ユーザーホームのパスが見つかりました: ${props.PATH}`,
    },
};
function reportIfFoundHomedir({ source, options, context, t, getUserHomeDirPattern, }) {
    const homedirPattern = getUserHomeDirPattern();
    const results = source.content.matchAll(homedirPattern);
    for (const result of results) {
        const index = result.index || 0;
        const match = result[0] || "";
        const range = [index, index + match.length];
        const allowedResults = matchPatterns(match, options.allows);
        if (allowedResults.length > 0) {
            continue;
        }
        context.report({
            message: t("HOMEDIR", {
                PATH: match,
            }),
            range,
        });
    }
}
export const creator = {
    messages,
    meta: {
        id: "@secretlint/secretlint-rule-no-homedir",
        recommended: true,
        type: "scanner",
        supportedContentTypes: ["text"],
        docs: {
            url: "https://github.com/secretlint/secretlint/blob/master/packages/%40secretlint/secretlint-rule-no-homedir/README.md",
        },
    },
    create(context, options) {
        const t = context.createTranslator(messages);
        const normalizedOptions = {
            allows: options.allows || [],
        };
        const userHomeDirPattern = multiplatformPath(options._debugHomeDir ?? os.homedir());
        /**
         * Clone RegExp to avoid reuse lastIndex
         */
        const getUserHomeDirPattern = () => {
            return new RegExp(userHomeDirPattern.source, userHomeDirPattern.flags);
        };
        return {
            file(source) {
                reportIfFoundHomedir({ source, options: normalizedOptions, context, t, getUserHomeDirPattern });
            },
        };
    },
};
